### ISSUE TITLE
NoSuchMethodError: method not found: 'get$span' on null

### ISSUE THREAD

**Issue by jsg2021** [sass/dart-sass#1204]
I've worked around all this, so there is no rush or pressure to act. I'm just adding documentation :)
https://github.com/webpack-contrib/sass-loader/issues/914

I believe there is an error in dart-sass when raising the deprecation warning in a Fiber ...or there is an error with toString implementation of the warning...
```
ModuleBuildError: Module build failed (from ../node_modules/sass-loader/dist/cjs.js):
SassError: NoSuchMethodError: method not found: 'get$span' on null
   ╷
25 │ $side-nav-link-bg-hover: hsla(0, 0, 0%, 0.025) !default;
   │                          ^^^^^^^^^^^^^^^^^^^^^
   ╵
``` 

Specifically the `NoSuchMethodError: method not found: 'get$span' on null`

_Originally posted by @jsg2021 in https://github.com/sass/dart-sass/issues/1203#issuecomment-759606271_

**Comment by nex3** [comment]
Can you provide some more details? What's the full stack trace of this error? Do you have a minimal reproduction of it?

**Comment by jsg2021** [comment]
I've been having trouble getting it to reproduce in smaller sample projects. I will keep trying after I get the full stack trace.

**Comment by jsg2021** [comment]
here is the full stack
```
Error: NoSuchMethodError: method not found: 'get$span' on null
   ╷
25 │ $side-nav-link-bg-hover: hsla(0, 0, 0, 0.025) !default;
   │                          ^^^^^^^^^^^^^^^^^^^^
   ╵
  ../../node_modules/@nti/foundation-sites-fork/scss/foundation/components/_side-nav.scss 25:26  @import
  ../../node_modules/@nti/foundation-sites-fork/scss/foundation.scss 31:9                        @import
  src/main/resources/scss/app.scss 7:9                                                           root stylesheet
    at Object._newRenderError (/home/jonathan/Workspace/nti/node_modules/sass/sass.dart.js:13494:19)
    at Object._wrapException (/home/jonathan/Workspace/nti/node_modules/sass/sass.dart.js:13337:16)
    at Object._renderSync (/home/jonathan/Workspace/nti/node_modules/sass/sass.dart.js:13312:18)
    at _render_closure.call$0 (/home/jonathan/Workspace/nti/node_modules/sass/sass.dart.js:79941:38)
    at Object.Primitives_applyFunction (/home/jonathan/Workspace/nti/node_modules/sass/sass.dart.js:1082:30)
    at Object.Function_apply (/home/jonathan/Workspace/nti/node_modules/sass/sass.dart.js:6003:16)
    at _callDartFunctionFast (/home/jonathan/Workspace/nti/node_modules/sass/sass.dart.js:7672:16)
    at /home/jonathan/Workspace/nti/node_modules/sass/sass.dart.js:7650:18
```

**Comment by jsg2021** [comment]
After tracing that stack, I cannot find the call site that is throwing. :disappointed: I can't seem to get the error when isolated to just sass/fibers ..or even with webpack/sass-loader added on. I can get it to happen every time in my large project.

**Comment by nex3** [comment]
Is your project public? If not, can you privately send me a copy? You can find my email in my profile.

**Comment by jsg2021** [comment]
> Is your project public? If not, can you privately send me a copy? You can find my email in my profile.

It's a company product, I can't. 😞  Could you use a vs code live share session to debug?

**Comment by jsg2021** [comment]
I finally got to the call site...
<img width="1440" alt="Screen Shot 2021-01-13 at 10 34 34 PM" src="https://user-images.githubusercontent.com/577453/104545404-95944900-55ef-11eb-845e-38616a7c9663.png">

**Comment by nex3** [comment]
Thanks for the extra detail! That definitely helps me figure out exactly where it's failling ([this line](https://github.com/sass/dart-sass/blob/6ea8ce6cf9c45a33f0735a4a8486b08af4c551d6/lib/src/visitor/evaluate.dart#L504)), but I'm still not sure *why*. `_callableNode` should never be null there—it's set [in `_runBuiltInCallable`](https://github.com/sass/dart-sass/blob/6ea8ce6cf9c45a33f0735a4a8486b08af4c551d6/lib/src/visitor/evaluate.dart#L2235) to `nodeWithSpan`, which comes [through `_runFunctionCallable`](https://github.com/sass/dart-sass/blob/6ea8ce6cf9c45a33f0735a4a8486b08af4c551d6/lib/src/visitor/evaluate.dart#L2235) [from `visitFunctionExpression`](https://github.com/sass/dart-sass/blob/6ea8ce6cf9c45a33f0735a4a8486b08af4c551d6/lib/src/visitor/evaluate.dart#L2083), where `node` is [clearly not null](https://github.com/sass/dart-sass/blob/6ea8ce6cf9c45a33f0735a4a8486b08af4c551d6/lib/src/visitor/evaluate.dart#L2058). Furthermore, there's nowhere `_calableNode` gets assigned *other than* `_runBuiltInCallable`, which always [reassigns it to its old value](https://github.com/sass/dart-sass/blob/6ea8ce6cf9c45a33f0735a4a8486b08af4c551d6/lib/src/visitor/evaluate.dart#L2295) when it exits, so *even if* another function call was happening somehow (which it shouldn't be) there's no way it could be clobbered between the initial assignment and the warning being emitted.

Without a way to reproduce this locally, we're going to have to rely on your debugging skills. Can you see if you can trace the state of `_callbackNode` (`_evaluate0$_callableNode` in the JS output) immediately prior to the warning being emitted to figure out how it's becoming null?

**Comment by jsg2021** [comment]
I will try! 😅

**Comment by jsg2021** [comment]
<img width="1440" alt="Screen Shot 2021-01-14 at 11 08 29 PM" src="https://user-images.githubusercontent.com/577453/104684409-9b0c9480-56be-11eb-9122-c62504d5a401.png">

<img width="1440" alt="Screen Shot 2021-01-14 at 11 14 30 PM" src="https://user-images.githubusercontent.com/577453/104684432-aa8bdd80-56be-11eb-9214-8e3a3ddf3801.png">

The `$.Zone__current.$index(0, C.Symbol__warn).$this` came from inspecting the warn:
<img width="593" alt="Screen Shot 2021-01-14 at 11 05 04 PM" src="https://user-images.githubusercontent.com/577453/104684359-7b756c00-56be-11eb-9a1e-cdecad739bea.png">


It looks like were are using two different "this" references... so the `_evaluate0$_callableNode` on the object `_this` is set, but the reference in `$.Zone__current.$index(0, C.Symbol__warn).$this` is not the same as `_this`...and its `_evaluate0$_callableNode` is null. With fibers set false, this goes down a completely different code path... 😅 I was going to try to compare... but oh well.

Any more tips? is the value pulled from Zone__current supposed to be the same as the value in _this? (from the pictures)

**Comment by nex3** [comment]
This gets really complex because of the way dart2js compiles classes and closures. `$.Zone__current.$index(0, C.Symbol__warn)` refers to [this callback](https://github.com/sass/dart-sass/blob/6ea8ce6cf9c45a33f0735a4a8486b08af4c551d6/lib/src/visitor/evaluate.dart#L503-L505), but you're right that it's strange that its `$this` field seems to be a different instance of `_EvaluateVisitor` than `_this`, and that only one of them seems to have `_callableNode` set. This suggests that maybe there are two overlapping instances of the Sass compiler running at once, and one of them is clobbering the other's warn callback. That callback is global-esque using a Dart-specific technique known as "zones" which *should* be safe across multiple concurrent invocations, but Fibers make asynchronous concurrency weird and that may be causing this inconsistency.

**Comment by nex3** [comment]
It looks like that was the root of the problem. I was able to reproduce a related bug locally, and I found a fix (#1208) that should address your issue as well. Thanks for the help debugging!

**Comment by jsg2021** [comment]
no, thank you! :)
